client-src/elements/chromedash-typeahead.js

```diff
@@ -0,0 +1,411 @@
+import {LitElement, css, html} from 'lit';
+import {ref, createRef} from 'lit/directives/ref.js';
+import {live} from 'lit/directives/live.js';
+import {SHARED_STYLES} from '../css/shared-css.js';
+import SlDropdown from '@shoelace-style/shoelace/dist/components/dropdown/dropdown.js';
+
+/* This file consists of 3 classes that together implement a "typeahead"
+   text field with autocomplete:
+
+   1. Chromedash-typeahead represents the overall UI widget, accepts a
+   `vocabulary` list of words, exposes a `value` string, and emits a
+   `sl-change` event when the user hits enter to submit the value.
+   Internally, it is responible for narrowing the vacabulary down to a list
+   of candidates based on the prefix that the user has typed.
+
+   2. Private class ChromedashTypeaheadDropdown subclasses SlDropdown and
+   removes code that would change the keyboard focus.
+
+   3. Private class ChromedashTypeaheadItem renders a single item in the
+   typeahead menu.  We do not use SlMenuItem because it steals keyboard focus.
+*/
+
+export class ChromedashTypeahead extends LitElement {
+  slDropdownRef = createRef();
+  slInputRef = createRef();
+
+  static get properties() {
+    return {
+      value: {type: String},
+      placeholder: {type: String},
+      candidates: {type: Array},
+      narrowCandidates: {attribute: false},
```

Comment:
type definition for these properties?

Reply:
These are not types in terms of compiler type checking, but they are a configuration of how lit converts from attribute string values into component properties.    I actually was not using `narrowCandidates` so I removed it.  Instead I use `voacabulary` for the entire list of known words, and `candidates` for the narrowed list.